godot-core/src/builtin/script.rs

```diff
@@ -0,0 +1,777 @@
+        // also guarantees that both vector length and capacity are equal.
+        let _drop = transfer_ptr_list_from_godot(p_prop_info, length);
+    }
+
+    /// # Safety
+    ///
+    /// - The engine has to provide a valid pointer to the instance data.
+    /// - `p_method` has to be a valid Godot string name ptr.
+    /// - `p_args` has to point to a list of Variant pointers of length `p_argument_count`.
+    /// - `r_return` has to be a valid variant pointer into which the return value can be moved.
+    /// - `r_error` has to point to a valid [`GDExtenstionCallError`] which can be modified to reflect the outcome of the method call.
+    pub(super) unsafe extern "C" fn call_func<T: ScriptInstance>(
+        p_self: sys::GDExtensionScriptInstanceDataPtr,
+        p_method: sys::GDExtensionConstStringNamePtr,
+        p_args: *const sys::GDExtensionConstVariantPtr,
+        p_argument_count: sys::GDExtensionInt,
+        r_return: sys::GDExtensionVariantPtr,
+        r_error: *mut sys::GDExtensionCallError,
+    ) {
+        let method = transfer_string_name_from_godot(p_method);
+        let args = Variant::unbounded_refs_from_sys(p_args, p_argument_count as usize);
+        let ctx = || format!("error when calling {}::call", type_name::<T>());
+
+        let result = handle_panic(ctx, || {
+            let instance = instance_data_as_script_instance::<T>(p_self);
+            borrow_instance_mut(instance).call(method.clone(), args)
+        });
+
+        match result {
+            Some(Ok(ret)) => {
+                transfer_variant_to_godot(ret, r_return);
+                (*r_error).error = sys::GDEXTENSION_CALL_OK;
+            }
+
+            Some(Err(err)) => (*r_error).error = err,
+            None => {
+                (*r_error).error = sys::GDEXTENSION_CALL_ERROR_INVALID_METHOD;
+            }
+        };
+    }
+
+    /// # Safety
+    ///
+    /// - The engine has to provide a valid pointer to the instance data.
+    /// - A pointer to a `Script` reference will be returned. The caller is then responsible for freeing the reference, including the adjustment
+    ///   of the reference count.
+    pub(super) unsafe extern "C" fn get_script_func<T: ScriptInstance>(
+        p_instance: sys::GDExtensionScriptInstanceDataPtr,
+    ) -> sys::GDExtensionObjectPtr {
+        let ctx = || format!("error when calling {}::get_script", type_name::<T>());
+
+        let script = handle_panic(ctx, || {
+            let instance = instance_data_as_script_instance::<T>(p_instance);
+
+            borrow_instance(instance).get_script().to_owned()
+        });
+
+        match script {
+            Some(script) => transfer_script_to_godot(&script),
+            None => std::ptr::null_mut(),
+        }
+    }
+
+    /// # Safety
+    ///
+    /// - The engine has to provide a valid pointer to the instance data.
+    /// - The boolean result is returned as an u8.
```

Comment:
Should this actually be a `u8` or a `sys::GDExtensionBool`? (same question below)